pkgs/by-name/ni/nixos-jobserver/nixos-jobserver.cpp

```diff
@@ -0,0 +1,463 @@
+ * this works fine for make, where one job does not substantially outlive its first
+ * failing component. it is not suitable for global jobservers, since a job may
+ * have failed without being able to return its tokens (eg because it was killed
+ * by SIGKILL, as nix-daemon does). we thus have to track how many tokens a client
+ * has acquired and return them if the client goes away, but to do this we also
+ * need to know when a client dies. this can be done with FUSE as done here, but
+ * could also be achieved with simple pipes or unix domain sockets. neither of
+ * these are a good choice for this problem:
+ *
+ * pipes are a resource shared between all processes that have opened said pipe.
+ * to be useful for a global jobserver we'd have to provide a unique pipe for
+ * each client, which requires more invasive changes to a nix configuration than
+ * using FUSE.
+ *
+ * both pipes and sockets have buffers. it is not possible to poll a pipe or
+ * socket for "the other end wants to read a byte", it is only possible to poll
+ * for available buffer space. distributing tokens among multiple clients of the
+ * jobserver must then be done by partitioning the available tokens and giving
+ * one partition to each client and periodically rebalancing according to actual
+ * usage. this periodic process necessarily causes more overhead than using the
+ * kernel as a notification source for "the other end wants to read a byte" and
+ * introduces unnecessary latency into token transfers between jobs.
+ */
+
+#define FUSE_USE_VERSION 35
+
+#include <errno.h>
+#include <fuse.h>
+#include <fuse_lowlevel.h>
+#include <grp.h>
+#include <poll.h>
+#include <pwd.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <sys/sysinfo.h>
+#include <sys/xattr.h>
+#include <unistd.h>
+
+#include <algorithm>
+#include <atomic>
+#include <iostream>
+#include <limits>
+#include <list>
+#include <memory>
+#include <optional>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+// 0x0A is ^J :)
+#define JOBSRV_IOC_SET_LIMIT _IOW(0x0A, 0, unsigned)
+
+struct PollHandleDestroy {
+    void operator()(fuse_pollhandle* ph) { fuse_pollhandle_destroy(ph); }
+};
+
+using PollHandle = std::unique_ptr<fuse_pollhandle, PollHandleDestroy>;
+
+struct Context {
+    unsigned tokens_held{0};
+    unsigned limit{std::numeric_limits<unsigned>::max()};
+    PollHandle poll;
+};
+
+struct stat jobserver_st = {
+    .st_ino = FUSE_ROOT_ID,
+    .st_mode = S_IFREG | 0660,
+    .st_uid = getuid(),
+    .st_gid = getgid(),
+};
+static const std::string acl_name = "system.posix_acl_access";
+static std::optional<std::vector<char>> acl;
+
+static unsigned tokens = 0;
+static std::unordered_set<Context*> poll_waiters;
+static std::unordered_map<fuse_req_t, Context*> read_waiters;
+
+static void wake_waiters()
+{
```

Comment:
contemplating how this works under contention:

1. clients doing blocking read() will always go first
2. this looks very liable to cause spurious wakeups on poll waiters since it wakes every poller so they will thundering-herd into read().

I am assuming both of these are by design, but I do wonder if poll/async io based users would get more or less totally deprioritized by accident over blocking ones: I am assuming that if you have a loop that looks like:

```
for (;;) {
    set = poll()
    if (set[fd] is readable) {
        toks = read(fd); 
        if (toks != -1) {
             // do whatever
             break;
        } else {
             if (errno == EAGAIN) {
                 continue;
             }
        }
    }
}
```

then if you have another process that wins the race from `poll()` to `read()`, some nonblocking client would not get tokens in a pretty unfairly distributed manner.

Is this a correct assessment?

To be fair, I don't think this could be written in a way that doesn't have this problem if it is FUSE. The only way I could conceive it working is if we used unix sockets that have *actual requests* for slots and somewhat of client identity (and then can also do requesting linker slots), then we could *deterministically* wake up single clients and give them the slot they asked for.

Reply:
could probably use a comment that says this *does* cause a thundering herd and inadvertently prioritize blocking io due to api limitation. feels like the kind of operational problem *someone* might end up running into one day and should possibly be documented as known.